tests/model_inheritance/test_abstract_inheritance.py

```diff
@@ -447,3 +447,34 @@ def fields(model):
                 ("name", models.CharField),
             ],
         )
+
+    @isolate_apps('model_inheritance')
+    def test_permission_inheritance_abstract(self):
+        """
+        Abstract permissions with %(model_name)s interpolation should become concrete
+        """
+
+        class A(models.Model):
+            class Meta:
+                app_label = 'model_inheritance'
+                abstract = True
+                permissions = [
+                    ('check_%(model_name)s', 'Check %(verbose_name)s'),
+                ]
+
+        class B(A):
+            pass
+
+        class C(A):
+            class Meta(A.Meta):
+                app_label = 'model_inheritance'
+                permissions = A.Meta.permissions + [
+                    ('concrete_c_permission', 'Concrete C permission'),
+                ]
+
+        self.assertEqual(A._meta.permissions, [('check_a', 'Check a')])
+        self.assertEqual(B._meta.permissions, [('check_b', 'Check b')])
+        self.assertEqual(
+            C._meta.permissions,
+            [('check_c', 'Check c'), ('concrete_c_permission', 'Concrete C permission')]
+        )
```

Comment:
I confirmed with a sample project that it's still affected by https://github.com/django/django/pull/6861#issuecomment-240427426. For example:
```python
class A(models.Model):
    class Meta:
        abstract = True
        verbose_name = "===custom abstract A===="
        permissions = [
            ('check_%(model_name)s', 'Check %(verbose_name)s'),
        ]

class B(A):
    pass


class C(A):
    class Meta(A.Meta):
        permissions = A.Meta.permissions + [
            ('concrete_c_permission', 'Concrete C permission'),
        ]
        verbose_name = "--C NAME--"
```
generates the following permissions:
```
sqlite> select * from auth_permission where codename like 'check%' or codename like 'concrete%';
40|7|check_b|Check ===custom abstract A====
41|8|check_c|Check ===custom abstract A====
42|8|concrete_c_permission|Concrete C permission
```
```

Reply:
EDIT: OK, it works when `verbose_name` is added with `permissions`, but (of course) doesn't take into account `verbose_name` added later.

Reply:
Ok, did some tests yesterday with verbose names:

```
        class A(models.Model):
            class Meta:
                app_label = 'model_inheritance'
                verbose_name = 'custom a'
                abstract = True
                permissions = [
                    ('check_%(model_name)s', 'Check %(verbose_name)s'),
                ]

        class B(A):
            class Meta(A.Meta):
                verbose_name = 'custom b'

        class C(A):
            class Meta(A.Meta):
                app_label = 'model_inheritance'
                permissions = A.Meta.permissions + [
                    ('concrete_c_permission', 'Concrete C permission'),
                ]

        class D(C):
            class Meta(C.Meta):
                permissions = C.Meta.permissions

        self.assertEqual(A._meta.permissions, [('check_a', 'Check custom a')])
        self.assertEqual(B._meta.permissions, [('check_b', 'Check custom b')])

        # Please note that the verbose name will be inherited from the parent class when it has
        # been set explicitly.
        self.assertEqual(
            C._meta.permissions,
            [('check_c', 'Check custom a'), ('concrete_c_permission', 'Concrete C permission')]
        )

        self.assertEqual(
            D._meta.permissions, [('check_d', 'Check custom a')],
        )
```

The above tests pass. That `verbose_name` inheritance seems to be consistent with general model inheritance behavior.
Though, I’d have expected that `D._meta.permissions` would contain `('concrete_c_permission', 'Concrete C permission')` (in that case, inheriting a non-abstract model). @felixxm – does that look right to you?